### PR TITLE
Enable ring buffer descriptor heaps as a runtime decision

### PR DESCRIPTION
**Pull request by jenatali**

### CODE REVIEW

File: include/ImmediateContext.inl [hunk 1]
```diff
@@ -1078,6 +1084,12 @@ inline void ImmediateContext::PreDispatch() noexcept(false)
     PIXScopedEvent(GetGraphicsCommandList(), 0ull, L"PreDispatch");
     PreRender(COMMAND_LIST_TYPE::GRAPHICS);
 
+    if (m_bUseRingBufferDescriptorHeaps)
+    {
+        // Always dirty the state when using ring buffer heaps because we can't safely reuse tables in that case.
```
Comment (amarpMSFT) [inline include/ImmediateContext.inl:1089]: In the original this was above PreRender

**Review by clandrew — APPROVED** [review]

File: include/ImmediateContext.inl [hunk 2]
```diff
@@ -496,6 +496,12 @@ inline void ImmediateContext::PreDraw() noexcept(false)
     PIXScopedEvent(GetGraphicsCommandList(), 0ull, L"PreDraw");
     PreRender(COMMAND_LIST_TYPE::GRAPHICS);
 
+    if (m_bUseRingBufferDescriptorHeaps)
+    {
+        // Always dirty the state when using ring buffer heaps because we can't safely reuse tables in that case.
```
Comment (amarpMSFT) [inline include/ImmediateContext.inl:501]: In the original this was above PreRender

File: src/ImmediateContext.cpp [hunk 3]
```diff
@@ -545,16 +546,13 @@ void ImmediateContext::RollOverHeap(OnlineDescriptorHeap& Heap) noexcept(false)
     }
     else
     {
-        if constexpr (g_bUseRingBufferDescriptorHeaps)
```
Comment (amarpMSFT) [inline src/ImmediateContext.cpp:548]: This check isn't in the original.  Should it be there too?

File: include/ImmediateContext.inl (hunk 1 above)
Comment (jenatali) [reply include/ImmediateContext.inl:1089]: Doesn't matter - PreRender only updates predication state, which isn't affected by these dirty flags. This was more aesthetic in my mind.

File: include/ImmediateContext.inl (hunk 2 above)
Comment (jenatali) [reply include/ImmediateContext.inl:501]: Same as above.

File: src/ImmediateContext.cpp (hunk 3 above)
Comment (jenatali) [reply src/ImmediateContext.cpp:548]: It absolutely should've been, since we were getting some telemetry spam, but that version isn't taking new updates.

**Review by alexgong14 — APPROVED** [review]

**Pull request merged**